src/lib.rs

```diff
@@ -195,3 +216,80 @@ where
         self.inner.ping()
     }
 }
+
+pub struct DTraceTransactionManager<C> {
+    _data: std::marker::PhantomData<C>,
+}
+
+impl<C> DTraceTransactionManager<C>
+where
+    C: Connection<TransactionManager = AnsiTransactionManager>,
+{
+    /// Compute the current transaction depth for the DTrace probes.
+    fn depth(conn: &mut DTraceConnection<C>) -> i64 {
+        let status = AnsiTransactionManager::transaction_manager_status_mut(&mut conn.inner);
+        match status.transaction_depth() {
+            Ok(Some(depth)) => i64::from(depth.get()),
+            Ok(None) => 0,
+            Err(_) => -1,
+        }
+    }
+}
+
+impl<C> TransactionManager<DTraceConnection<C>> for DTraceTransactionManager<C>
+where
+    C: Connection<TransactionManager = AnsiTransactionManager>,
+    C::Backend: Default,
+    <C::Backend as Backend>::QueryBuilder: Default,
+{
+    type TransactionStateData = AnsiTransactionManager;
+
+    fn begin_transaction(conn: &mut DTraceConnection<C>) -> QueryResult<()> {
+        // TODO-performance: We're unconditionally computing the transaction
+        // depth here, even if the probe is not enabled.
+        //
+        // This ultimately comes from the interaction of a few things. These
+        // trait methods don't make it easy to store state -- since they take
+        // the mutable connection, not `&mut self`, we have to store everything
+        // on the connection type, but that interacts with the
+        // `Connection::transaction_state()` method weirdly. Second, even if we
+        // could do that, there's no good way to keep that in sync if the probes
+        // are disabled _while_ a transaction is outstanding (we'd end up
+        // thinking we were still in a transaction, when we're not anymore).
+        //
+        // Last, this interacts with a `Clone` bound on the `usdt` crate's
```

Comment:
I wonder if there's a model of this where we can use an `Arc` or `Rc` to pass this value, get mutable access to it in the probe closure, and then get mutable access again when we later call `AnsiTransactionManager::begin_transaction`.

Regardless, I agree with your sentiment, this probably isn't worth doing now, since the cost should be relatively small anyway.

Reply:
Oh, that's an interesting idea, thanks. We're hoping to get rid of the bound completely, but not sure the best way to do that.